src/core/varlink-metrics.c

```diff
@@ -0,0 +1,306 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+#include "hashmap.h"
+#include "install.h"
+#include "json-util.h"
+#include "manager.h"
+#include "metrics.h"
+#include "set.h"
+#include "sd-json.h"
+#include "sd-varlink.h"
+#include "unit.h"
+#include "unit-def.h"
+#include "varlink-metrics.h"
+
+static int unit_states_total_build_json_one(
+               sd_varlink *link,
+               UnitActiveState state,
+               unsigned count,
+               bool more) {
+       int r;
+       _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+
+
+       assert(link);
+
+
+       r = METRIC_JSON_BUILD_UNSIGNED(
+                       &v,
+                       METRIC_IO_SYSTEMD_MANAGER_UNITS_BY_STATE_TOTAL,
+                       /* object= */ NULL,
+                       count,
+                       /* fields */
+                       "state",
+                       unit_active_state_to_string(state));
+       if (r < 0)
+               return r;
+
+
+       if (more)
+               return sd_varlink_notify(link, v);
+
+
+       return sd_varlink_reply(link, v);
+}
+
+static int unit_types_total_build_json_one(sd_varlink *link, Manager *manager, UnitType *type, bool more) {
+       int r;
+       unsigned count = 0;
+       _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+
+
+       assert(link);
+       assert(manager);
+       assert(type);
+
+
+       // TODO: This needs a rework/improvement
+       LIST_FOREACH(units_by_type, _u, manager->units_by_type[*type])
+               count++;
+
+
+       r = METRIC_JSON_BUILD_UNSIGNED(
+               &v,
+               METRIC_IO_SYSTEMD_MANAGER_UNITS_BY_TYPE_TOTAL,
+               /* object= */ NULL,
+               count,
+               /* fields */
+               "type", unit_type_to_string(*type));
+       if (r < 0)
+               return r;
+
+
+       if (more)
+               return sd_varlink_notify(link, v);
+
+
+       return sd_varlink_reply(link, v);
+}
+
+static int unit_types_total_build_json(sd_varlink *link, void *userdata, bool more) {
+       int r;
+       Manager *manager = ASSERT_PTR(userdata);
+
+       assert(link);
+       assert(manager);
+
+       UnitType *t, *previous_type = NULL;
+       UnitType type;
+       for (int i = 0; i < _UNIT_TYPE_MAX; i++) {
+               type = (UnitType) i;
+               t = &type;
+
+               if (previous_type) {
+                       r = unit_types_total_build_json_one(link, manager, previous_type, more);
+                       if (r < 0)
+                               return r;
+               }
+
+               previous_type = t;
+      }
+
+       if (previous_type) {
+               r = unit_types_total_build_json_one(link, manager, previous_type, more);
+               if (r < 0)
+                       return r;
+       } else
+               return sd_varlink_error(link, VARLINK_ERROR_METRICS_NO_SUCH_METRIC, NULL);
```

Comment:
is this if condition even possible? I don't see a case where previous_type would be NULL.
Also, I don't think we should ever return `VARLINK_ERROR_METRICS_NO_SUCH_METRIC` from a callback.

`vl_method_list()` can return it though as result of the filtering (which is not implemented yet)